### PR TITLE
[MCC-288584] Feature/bulk upload new base

### PR DESCRIPTION
**Pull request by fpcyan**
@mdsol/team04 

Not sure why these incredibly old commits are showing up, but these changes look like they're in 0.1.1-stable.

This updates the gem to include buffering. You may find it familiar as it's inspired by the persistence buffer in mdsol/the_policy_machine.

Buffering provides two notable benefits over immediate enqueuing of messages:
1. Fewer calls to SQS queues can improve latency in situations of high volume.
2. Enqueuing can be done transactionally, so even if the transaction rolls back, that doesn't necessarily mean the message has done out.

### CODE REVIEW

File: lib/delayed/backend/sqs.rb [hunk 1]
```diff
@@ -85,20 +113,45 @@ def save
...
         end
 
-        # Destroy the job; that is, delete it from the SQS queue and don't save it anywhere.
+        def send_later(message)
+          buffer[@queue_name] = [[]] unless buffer[@queue_name]
+          current_buffer = buffer[@queue_name]
+
+          current_buffer_size = current_buffer.last.reduce(0) { |m, msg| m + msg[:message_body].bytesize }
+          if current_buffer_size + message[:message_body].bytesize >= ::DelayedJobSqs::Document::MAX_SQS_MESSAGE_SIZE_IN_BYTES ||
+             current_buffer.last.size >= 10
```
Comment (HonoreDB) [inline lib/delayed/backend/sqs.rb]: Is 10 a max number of messages per batch in the SQS API?

Comment (DMcKinnon-mdsol) [reply lib/delayed/backend/sqs.rb]: Womp

File: lib/delayed/backend/sqs.rb [hunk 2]
```diff
@@ -85,20 +113,45 @@ def save
...
           save
         end
 
-        # Destroy the job; that is, delete it from the SQS queue and don't save it anywhere.
+        def send_later(message)
+          buffer[@queue_name] = [[]] unless buffer[@queue_name]
+          current_buffer = buffer[@queue_name]
+
+          current_buffer_size = current_buffer.last.reduce(0) { |m, msg| m + msg[:message_body].bytesize }
+          if current_buffer_size + message[:message_body].bytesize >= ::DelayedJobSqs::Document::MAX_SQS_MESSAGE_SIZE_IN_BYTES ||
```
Comment (DMcKinnon-mdsol) [inline lib/delayed/backend/sqs.rb]: I don't love multiline if's. Is it worth just separating this out into its own ? method?

**Comment by HonoreDB** [comment]
Very cool! Mind adding instructions to the README as well?

File: lib/delayed/backend/sqs.rb (hunk 1 above)
Comment (fpcyan) [reply lib/delayed/backend/sqs.rb]: @HonoreDB 256 KB or 10 messages per batch

File: lib/delayed/backend/sqs.rb (hunk 2 above)
Comment (fpcyan) [reply lib/delayed/backend/sqs.rb]: I feel like it would still be too long, but I would happily do things to make this smaller.

Comment (jfeltesse-mdsol) [reply lib/delayed/backend/sqs.rb]: and also the line is > 120 chars.

File: lib/delayed/backend/sqs.rb [hunk 3]
```diff
@@ -85,20 +113,45 @@ def save
...
-          # Resend the message before deleting from queue to ensure if resend fails, message stays and job can be retried when visibility timeout is exceeded
-          sqs.queues.named(queue_name).send_message(payload, :delay_seconds  => @delay)
-          @msg.delete if @msg # TODO:  potential problem here in that there may be multiple copies of this message on the q since SQS guarantees to write at least once
-          
+          @msg.delete if @msg
+
+          maxed_delay = [900, @delay + 5 + attempts ** 4].min
+
+          if buffering?
+            send_later({ message_body: payload, delay_seconds: maxed_delay })
```
Comment (jfeltesse-mdsol) [inline lib/delayed/backend/sqs.rb]: unneccessary `{ }`

File: lib/delayed/serialization/sqs.rb [hunk 4]
```diff
@@ -1,6 +1,38 @@
...
+require 'base64'
+
 module DelayedJobSqs
   module Document
+    MAX_SQS_MESSAGE_SIZE_IN_BYTES = 2**18
+
+    def self.sqs_safe_json_dump(obj)
+      json = JSON.dump(obj)
+      if json.bytesize >= MAX_SQS_MESSAGE_SIZE_IN_BYTES
+        JSON.dump( dj_compressed_document: compress(json) )
```
Comment (jfeltesse-mdsol) [inline lib/delayed/serialization/sqs.rb]: spaces within `( )`

File: lib/delayed/backend/sqs.rb (hunk 1 above)
Comment (jfeltesse-mdsol) [reply lib/delayed/backend/sqs.rb]: this should be a constant then, me thinks.

File: lib/delayed/backend/sqs.rb [hunk 5]
```diff
@@ -85,20 +113,45 @@ def save
...
+          else
+            sqs.queues.named(queue_name).send_message(payload, :delay_seconds  => @delay )
```
Comment (jfeltesse-mdsol) [inline lib/delayed/backend/sqs.rb]: old hash syntax

File: lib/delayed/backend/sqs.rb [hunk 6]
```diff
@@ -85,20 +113,45 @@ def save
...
-          payload = JSON.dump(@attributes)
+          payload = ::DelayedJobSqs::Document.sqs_safe_json_dump(@attributes)
 
-          # Resend the message before deleting from queue to ensure if resend fails, message stays and job can be retried when visibility timeout is exceeded
-          sqs.queues.named(queue_name).send_message(payload, :delay_seconds  => @delay)
-          @msg.delete if @msg # TODO:  potential problem here in that there may be multiple copies of this message on the q since SQS guarantees to write at least once
-          
+          @msg.delete if @msg
+
+          maxed_delay = [900, @delay + 5 + attempts ** 4].min
```
Comment (jfeltesse-mdsol) [inline lib/delayed/backend/sqs.rb]: how about having a comment putting this in words?

File: lib/delayed/backend/sqs.rb [hunk 7]
```diff
@@ -16,13 +16,41 @@ class Job
...
+          @buffering = false
+        end
+
+        def self.clear_buffer!
+          @buffer = nil
+        end
+
+        def self.buffer
+          @buffer ||= {}
+        end
```
Comment (jfeltesse-mdsol) [inline lib/delayed/backend/sqs.rb]: IMO those `self.xxxxx` methods should be grouped within a `class << self` block.

**Review by jfeltesse-mdsol — COMMENTED** [review]
mostly cosmetic comments for me

File: lib/delayed/backend/sqs.rb (hunk 7 above)
Comment (fpcyan) [reply lib/delayed/backend/sqs.rb]: Isn't `class << self` old syntax?

File: lib/delayed/backend/sqs.rb (hunk 6 above)
Comment (fpcyan) [reply lib/delayed/backend/sqs.rb]: I didn't write this particular thing.

Comment (fpcyan) [reply lib/delayed/backend/sqs.rb]: but I added a thing anyway haha

**Comment by fpcyan** [comment]
@mdsol/team04 all comments addressed.

File: lib/delayed/backend/sqs.rb (hunk 7 above)
Comment (jfeltesse-mdsol) [reply lib/delayed/backend/sqs.rb]: not at all. It's alive and well and in such cases comes in handy to save you from prefixing all your methods with `self.`

**Comment by DMcKinnon-mdsol** [comment]
Thanks @fpcyan! This PR looks good to me.

Because this is a new feature (instead of just a bug fix), shouldn't PR be merged into develop and then we'd create a new stable branch (or tag, preferably) off develop > master?

**Comment by fpcyan** [comment]
@DMcKinnon-mdsol develop looks really weird, and I didn't want to be the one to make it less weird while also making significant changes myself. I would prefer to merge the version branch into develop after the fact. Same argument against using a tag for this.

**Comment by DMcKinnon-mdsol** [comment]
Hmm. I see what you're saying about not wanting to fix develop in this PR, but we shouldn't update an existing stable branch with a new feature. How do you feel about making a new stable branch off this stable branch and retarget this PR there?

**Comment by fpcyan** [comment]
`0.2.0-stable` was created by me by creating a fresh branch off of `0.1.1-stable`. So, I agree with your opinion here, and have preemptively done just that!

**Comment by DMcKinnon-mdsol** [comment]
Ha perfect!

**Pull request merged**